src/goto-programs/destructor_tree.h

```diff
@@ -0,0 +1,101 @@
+// Copyright 2016-2018 Diffblue Limited.
+
+#ifndef CPROVER_GOTO_PROGRAMS_DESTRUCTOR_TREE_H
+#define CPROVER_GOTO_PROGRAMS_DESTRUCTOR_TREE_H
+
+#include <util/graph.h>
+#include <util/std_code.h>
+
+typedef std::size_t node_indext;
+
+class ancestry_resultt
+{
+public:
+  explicit ancestry_resultt(node_indext node)
+    : common_ancestor(node),
+      left_pre_intersection_size(0),
+      right_pre_intersection_size(0)
+  {
+  }
+  ancestry_resultt(
+    node_indext node,
+    std::size_t left_pre_size,
+    std::size_t right_pre_size)
+    : common_ancestor(node),
+      left_pre_intersection_size(left_pre_size),
+      right_pre_intersection_size(right_pre_size)
+  {
+  }
+  node_indext common_ancestor;
+  std::size_t left_pre_intersection_size;
+  std::size_t right_pre_intersection_size;
+};
+
+/// Tree to keep track of the destructors generated along each branch of a
+/// function. Can be used to compare and find out what dead
+/// instructions are needed when moving from one branch to another.
+class destructor_treet
+{
+public:
+  destructor_treet()
+  {
+    // We add a default node to the graph to act as a root for path traversal.
+    this->destruction_graph.add_node();
```

Comment:
It's still not clear to me: you test for `destruction_graph.empty()` in some places - how is that possible?

Reply:
Ah, sorry. Originally there was a version which didn't have a root node (also the reason why there were a lot of optionals). This was a leftover of that and should be removed.